linfa-clustering/src/gaussian_mixture/algorithm.rs

```diff
@@ -0,0 +1,494 @@
+        }
+    }
+}
+
+impl<F: Float + Into<f64>> GaussianMixtureModel<F> {
+    pub fn params(n_clusters: usize) -> GmmHyperParamsBuilder<F, Isaac64Rng> {
+        GmmHyperParams::new(n_clusters)
+    }
+
+    pub fn params_with_rng<R: Rng + Clone>(
+        n_clusters: usize,
+        rng: R,
+    ) -> GmmHyperParamsBuilder<F, R> {
+        GmmHyperParams::new_with_rng(n_clusters, rng)
+    }
+
+    pub fn weights(&self) -> &Array1<F> {
+        &self.weights
+    }
+
+    pub fn means(&self) -> &Array2<F> {
+        &self.means
+    }
+
+    pub fn precisions(&self) -> &Array3<F> {
+        &self.precisions
+    }
+
+    pub fn centroids(&self) -> &Array2<F> {
+        self.means()
+    }
+
+    fn new<D: Data<Elem = F>, R: Rng + Clone, T: Targets>(
+        hyperparameters: &GmmHyperParams<F, R>,
+        dataset: &Dataset<ArrayBase<D, Ix2>, T>,
+        mut rng: R,
+    ) -> GaussianMixtureModel<F> {
+        let observations = dataset.records().view();
+        let n_samples = observations.nrows();
+
+        let resp = match hyperparameters.init_method() {
+            GmmInitMethod::KMeans => {
+                let model = KMeans::params_with_rng(hyperparameters.n_clusters(), rng)
+                    .build()
+                    .fit(&dataset);
+                let mut resp = Array::<F, Ix2>::zeros((n_samples, hyperparameters.n_clusters()));
+                for (k, idx) in model.predict(dataset.records()).iter().enumerate() {
+                    resp[[k, *idx]] = F::from(1.).unwrap();
+                }
+                resp
+            }
+            GmmInitMethod::Random => {
+                let mut resp = Array2::<f64>::random_using(
+                    (n_samples, hyperparameters.n_clusters()),
+                    Uniform::new(0., 1.),
+                    &mut rng,
+                );
+                let totals = &resp.sum_axis(Axis(1)).insert_axis(Axis(0));
+                resp = (resp.reversed_axes() / totals).reversed_axes();
+                resp.mapv(|v| F::from(v).unwrap())
+            }
+        };
+
+        let (mut weights, means, covariances) = Self::estimate_gaussian_parameters(
+            &observations,
+            &resp,
+            hyperparameters.covariance_type(),
+            hyperparameters.reg_covariance(),
+        );
+        weights = weights / F::from(n_samples).unwrap();
+
+        // GmmCovarType = full
+        let precisions = Self::compute_precision_cholesky_full(&covariances);
+
+        GaussianMixtureModel {
+            covar_type: *hyperparameters.covariance_type(),
+            weights,
+            means,
+            precisions,
+        }
+    }
+
+    fn estimate_gaussian_parameters<D: Data<Elem = F>>(
+        observations: &ArrayBase<D, Ix2>,
+        resp: &Array2<F>,
+        _covar_type: &GmmCovarType,
+        reg_covar: F,
+    ) -> (Array1<F>, Array2<F>, Array3<F>) {
+        let nk = resp.sum_axis(Axis(0)) + F::from(10.).unwrap() * F::epsilon();
```

Comment:
you could also return an error when one of the clusters becomes empty

Reply:
ok